File: 0x0092_146.LRU_Cache/solution.py
class DLN:
    def __init__(self, k, v):
        self.k = k
        self.v = v
        self.p = None
        self.n = None

    def rm(self):
        p, n = self.p, self.n
        if self.p:
            self.p.n = self.n
        if self.n:
            self.n.p = self.p
        self.p = None
        self.n = None
        return p, n


class LRUCache:
    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.size = 0
        self.capacity = capacity
        self.head = None
        self.tail = None
        self.k2n = {}

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.k2n:
            return -1
        n = self.k2n[key]
        out = n.v
        if n == self.head:
            return out
        self._remove(key)
        self._add(key, out)
        return out

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        if key in self.k2n:
            self._remove(key)
        self._add(key, value)
        if self.size > self.capacity:
            self._evict()

    def _remove(self, key):
        node = self.k2n.pop(key)
        p, n = node.rm()
        if node == self.head:
            self.head = n
        if node == self.tail:
            self.tail = p
        self.size -= 1

    def _add(self, key, value):
        node = DLN(key, value)
        self.k2n[key] = node
        if self.head:
            old_head = self.head
            self.head = node
            node.n = old_head
            old_head.p = node
        else:
            self.head = node
            self.tail = node
        self.size += 1

    def _evict(self):
        node = self.tail
        self.k2n.pop(node.k)
        self.tail = node.p
        node.rm()
        self.size -= 1


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0x0092_146.LRU_Cache/solution.java
class LRUCache {
    class DLinkedNode {
        int key;
        int val;
        DLinkedNode next;
        DLinkedNode prev;
        public DLinkedNode(int key1, int val1) {
            key=key1;
            val=val1;
            next=null;
            prev=null;
        }
    }
    
    class DLinkedList {
        DLinkedNode head;
        DLinkedNode tail;
        
        public DLinkedList() {
            head=null;
            tail=null;
        }
        
        // prepend a node
        public void prepend(DLinkedNode node) {
            if (head==null) {
                head=node;
                tail=node;
            } else {
                head.prev=node;
                node.next=head;
                head=node;
            }
        }
        
        // remove something found in the linkedlist
        public void remove(DLinkedNode node) {
            if (node==head && node == tail) {
                head=null;
                tail=null;
            } else if (node==head) {
                head=head.next;
                head.prev=null;
            } else if (node==tail) {
                tail=tail.prev;
                tail.next=null;
            } else {
                node.next.prev=node.prev;
                node.prev.next=node.next;
            }
        }
        
    }
    
    int cap;
    int length;
    Map<Integer,DLinkedNode> map;
    DLinkedList myList;
    
    public LRUCache(int capacity) {
        cap=capacity;
        length=0;
        map=new HashMap<Integer,DLinkedNode>();
        myList=new DLinkedList();
    }
    
    public int get(int key) {
        if (map.containsKey(key)) {
            DLinkedNode node =map.get(key);
            myList.remove(node);
            myList.prepend(node);  
            return node.val;
        } else {
            return -1;
        }
    }
    
    public void put(int key, int value) {
        if (map.containsKey(key)) {
            DLinkedNode node =map.get(key);
            node.val=value;
            myList.remove(node);
            myList.prepend(node);  
        } else {
            DLinkedNode node = new DLinkedNode(key,value);
            myList.prepend(node);
            map.put(key,node);
            
            if (length==cap) {
                map.remove(myList.tail.key);
                myList.remove(myList.tail);
            } else {
                length++;
            }
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */